source/application/AMPIC887DriverServer/AMPIC887Controller.cpp

```diff
@@ -61,10 +114,954 @@ QString AMPIC887Controller::status() const
+{
+	return controllerState_->hexapodState()->highSoftLimits().values(axes);
+}
+
+bool AMPIC887Controller::setHighSoftLimits(const AMPIC887AxisMap<double> &highSoftLimits)
+{
+	if(controllerState_->hexapodState()->highSoftLimits() == highSoftLimits) {
+		return true;
+	}
+
+	AMGCS2SetHighSoftLimitsCommand setHighSoftLimitCommand(highSoftLimits);
+	runCommand(&setHighSoftLimitCommand);
+
+	if(setHighSoftLimitCommand.runningState() != AMGCS2Command::Succeeded) {
+		setError(setHighSoftLimitCommand.lastError());
+		return false;
+	}
+
+	controllerState_->hexapodState()->setHighSoftLimits(highSoftLimits);
+	return true;
+}
+
+bool AMPIC887Controller::softLimitStatus(AMGCS2::Axis axis) const
+{
+	return controllerState_->hexapodState()->softLimitState(axis);
+}
+
+bool AMPIC887Controller::setSoftLimitStatus(AMGCS2::Axis axis, bool softLimitState)
+{
+	AMPIC887AxisMap<bool> singleValue;
+	singleValue.insert(axis, softLimitState);
+	return setSoftLimitStatuses(singleValue);
+}
+
+AMPIC887AxisMap<bool> AMPIC887Controller::softLimitStatuses(const AMPIC887AxisCollection &axes) const
+{
+	return controllerState_->hexapodState()->softLimitStates().values(axes);
+}
+
+bool AMPIC887Controller::setSoftLimitStatuses(const AMPIC887AxisMap<bool> &softLimitStates)
+{
+	if(controllerState_->hexapodState()->softLimitStates() == softLimitStates) {
+		return true;
+	}
+
+	AMGCS2SetSoftLimitsStatusCommand setSoftLimitStatesCommand(softLimitStates);
+	runCommand(&setSoftLimitStatesCommand);
+
+	if(setSoftLimitStatesCommand.runningState() != AMGCS2Command::Succeeded) {
+		setError(setSoftLimitStatesCommand.lastError());
+		return false;
+	}
+
+	controllerState_->hexapodState()->setSoftLimitStates(softLimitStates);
+	return true;
+}
+
+bool AMPIC887Controller::limitSwitchStatus(AMGCS2::Axis axis) const
+{
+	return controllerState_->hexapodState()->limitSwitchState(axis);
+}
+
+AMPIC887AxisMap<bool> AMPIC887Controller::limitSwitchStatuses(const AMPIC887AxisCollection &axes) const
+{
+	return controllerState_->hexapodState()->limitSwitchStates().values(axes);
+}
+
+double AMPIC887Controller::minCommandablePosition(AMGCS2::Axis axis) const
+{
+	return controllerState_->hexapodState()->minCommandablePosition(axis);
+}
+
+AMPIC887AxisMap<double> AMPIC887Controller::minCommandablePositions(const AMPIC887AxisCollection &axes) const
+{
+	return controllerState_->hexapodState()->minCommandablePositions().values(axes);
+}
+
+double AMPIC887Controller::maxCommandablePosition(AMGCS2::Axis axis) const
+{
+	return controllerState_->hexapodState()->maxCommandablePosition(axis);
+}
+
+AMPIC887AxisMap<double> AMPIC887Controller::maxCommandablePositions(const AMPIC887AxisCollection &axes) const
+{
+	return controllerState_->hexapodState()->maxCommandablePositions().values(axes);
+}
+
+bool AMPIC887Controller::onTargetState(AMGCS2::Axis axis)
+{
+	double currentAxisPosition = currentPosition(axis);
+	double currentTargetPosition = targetPosition(axis);
+
+	double epsilon = 0.0001;
```

Comment:
It worked in testing, though that was a while ago.

The thing with this value is that it represents a limit of certainty on the position of the axes from this point higher. The position will always be +/- this epsilon.

With the importance of that fact in mind I've altered these to be a 
``` C++
#define AXIS_POSITION_TOLERANCE 0.0001
```
in the controller class, and then changed all these vague references to epsilon to AXIS_POSITION_TOLERANCE. That way, when I later come to test this value to see how far I can push it I only ever have to change it in one place, and it should remain consistent across the whole system.